### ISSUE TITLE
panic: sync: WaitGroup is reused before previous Wait has returned

### ISSUE THREAD

**Issue by Renay** [docker/docker#28161]
**Steps to reproduce the issue:**
None

**Describe the results you received:**

daemons crashed. This is what I see on logs :

```
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.364614055Z" level=info msg="containerd: 368f199d569ef1594cbe66d53369f37c94ef26715d193d5eff5b91223ac93c49:8afb04e8fff33138a114a1e48452bd681abe3e72033a4c3d907fb34ddc165666 (pid 19979) has become an orphan, killing it"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.491048009Z" level=warning msg="containerd: unable to save ce609a558d20fa6494eb51486ec5ce15571cbc4ade750108175b26ea0a1a2267:5ac116b4ecaed32e9fccf4ab3b4ad436f36263f638647e588ab670fc5fdc054c starttime: open /proc/20001/stat: no such file or directory"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.491343165Z" level=info msg="containerd: ce609a558d20fa6494eb51486ec5ce15571cbc4ade750108175b26ea0a1a2267:5ac116b4ecaed32e9fccf4ab3b4ad436f36263f638647e588ab670fc5fdc054c (pid 20001) has become an orphan, killing it"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.557266677Z" level=warning msg="containerd: unable to save 7599c4fe01140050369e3526ee6f731db7a2c0beacb695052a360940345df70d:fd1447e456cba21f34b617e2376de3725c63ca719fc84a119f774476dd6dde52 starttime: open /proc/20020/stat: no such file or directory"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.557580385Z" level=info msg="containerd: 7599c4fe01140050369e3526ee6f731db7a2c0beacb695052a360940345df70d:fd1447e456cba21f34b617e2376de3725c63ca719fc84a119f774476dd6dde52 (pid 20020) has become an orphan, killing it"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.669930826Z" level=warning msg="containerd: unable to save f0bbdab0330ee236f0e8c3b55fa255793c03e66effc95422a100180819cdb644:2f32ac514d69ac7c9cf8d9b8ed1564362fe196e411cb2e7399d400cb5cb2a467 starttime: open /proc/20047/stat: no such file or directory"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.670188797Z" level=info msg="containerd: f0bbdab0330ee236f0e8c3b55fa255793c03e66effc95422a100180819cdb644:2f32ac514d69ac7c9cf8d9b8ed1564362fe196e411cb2e7399d400cb5cb2a467 (pid 20047) has become an orphan, killing it"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.760769154Z" level=warning msg="containerd: unable to save d9b4c86d9419706715e5f76771172d500468968f15df3d582b7f00887188672c:169516f1c66a8030092b9c401b080860684a53764fe8c735da836a6584bb7018 starttime: open /proc/20068/stat: no such file or directory"
Nov  8 09:59:50  dockerd[24383]: time="2016-11-08T09:59:50.761052184Z" level=info msg="containerd: d9b4c86d9419706715e5f76771172d500468968f15df3d582b7f00887188672c:169516f1c66a8030092b9c401b080860684a53764fe8c735da836a6584bb7018 (pid 20068) has become an orphan, killing it"

Nov  8 09:59:50  dockerd[24383]: panic: sync: WaitGroup is reused before previous Wait has returned 
Nov  8 09:59:50  dockerd[24383]: goroutine 1450730 [running]:
Nov  8 09:59:50  dockerd[24383]: panic(0x179c880, 0xc821d4c5e0)
Nov  8 09:59:50  dockerd[24383]: /usr/local/go/src/runtime/panic.go:481 +0x3e6
Nov  8 09:59:50  dockerd[24383]: sync.(*WaitGroup).Wait(0xc8245e5c80)
Nov  8 09:59:50  dockerd[24383]: /usr/local/go/src/sync/waitgroup.go:129 +0x114
Nov  8 09:59:50  dockerd[24383]: github.com/docker/docker/daemon.(*Daemon).StateChanged(0xc820406ea0, 0xc82179cf00, 0x40, 0x1e40700, 0xc, 0x0, 0xc82179d200, 0x40, 0x0, 0x0, ...)
Nov  8 09:59:50  dockerd[24383]: /go/src/github.com/docker/docker/daemon/monitor.go:68 +0xecc
Nov  8 09:59:50  dockerd[24383]: github.com/docker/docker/libcontainerd.(*container).handleEvent.func1()
Nov  8 09:59:50  dockerd[24383]: /go/src/github.com/docker/docker/libcontainerd/container_linux.go:184 +0xab
Nov  8 09:59:50  dockerd[24383]: github.com/docker/docker/libcontainerd.(*queue).append.func1(0xc82432fc01, 0xc823ec8de0, 0xc823c46360, 0xc821d6a2a0)
Nov  8 09:59:50  dockerd[24383]: /go/src/github.com/docker/docker/libcontainerd/queue_linux.go:26 +0x47
Nov  8 09:59:50  dockerd[24383]: created by github.com/docker/docker/libcontainerd.(*queue).append
Nov  8 09:59:50  dockerd[24383]: /go/src/github.com/docker/docker/libcontainerd/queue_linux.go:28 +0x1da

```
** Additional informations :**

I have a lot of logs like : 

Nov  8 01:44:05  dockerd[24383]: time="2016-11-08T01:44:05.137187495Z" level=info msg="containerd: 59b95c83cc24830c60d20231090f451721ec9c118a965d0815e698c693130543:4213194b565b0f581be1bb5cd347c4502f54aeb92f870165dc947842c721e909 (pid 21089) has become an orphan, killing it"

grep 'has become an orphan, killing it' /var/log/docker.log | wc -l
77229

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Fri Nov  4 13:16:53 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Fri Nov  4 13:16:53 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 122
 Running: 119
 Paused: 0
 Stopped: 3
Images: 10
Server Version: 1.12.3
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay host null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.2.0-ovh-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 57.6 GiB
Name: xxxxxx
ID: CMGS:5EE5:TMEG:OBPI:4VS4:KOXS:SZJT:CONL:5K3X:EWXO:FGQ6:UAVS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
Insecure Registries:
 127.0.0.0/8

```

**Comment by ripcurld00d** [comment]
@Renay did that happen when you tried to restart/stop the daemon?

Do you have an idea to reproduce these errors?

Thanks.

**Comment by Renay** [comment]
The daemon crashed. 

All my processes inside containers were attached to the main pid 1.

In order to relaunch docker, I have to kill all my processes and start the daemon.

I don't know how to reproduce the problem yet. It's maybe related to ceph storage behind but this problem never occured before ( just a feeling for now )

**Comment by thaJeztah** [comment]
ping @mlaventure @crosbymichael PTAL

@Renay I see you're running on a custom kernel, IIRC we've had issues with kernel configurations of OVH, just to be sure, could you also run the `check-config.sh` script; https://github.com/docker/docker/blob/master/contrib/check-config.sh to see if there's something standing out in the configuration?

**Comment by Renay** [comment]
```
./checkconfig.sh 
warning: /proc/config.gz does not exist, searching other paths for kernel config ...
info: reading kernel config from /boot/config-4.2.0-ovh-amd64 ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_NF_NAT_IPV4: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_IPVS: enabled (as module)
- CONFIG_IP_NF_NAT: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled

Optional Features:
- CONFIG_USER_NS: enabled
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: missing
- CONFIG_MEMCG_SWAP: enabled
- CONFIG_MEMCG_SWAP_ENABLED: enabled
- CONFIG_MEMCG_KMEM: missing
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_IOSCHED_CFQ: enabled
- CONFIG_CFQ_GROUP_IOSCHED: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: missing
- CONFIG_NET_CLS_CGROUP: enabled (as module)
- CONFIG_CGROUP_NET_PRIO: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: missing
- CONFIG_IP_VS: enabled (as module)
- CONFIG_IP_VS_NFCT: enabled
- CONFIG_IP_VS_RR: enabled (as module)
- CONFIG_EXT3_FS: missing
- CONFIG_EXT3_FS_XATTR: missing
- CONFIG_EXT3_FS_POSIX_ACL: missing
- CONFIG_EXT3_FS_SECURITY: missing
    (enable these ext3 configs if you are using ext3 as backing filesystem)
- CONFIG_EXT4_FS: enabled (as module)
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
      Optional (for encrypted networks):
      - CONFIG_CRYPTO: enabled
      - CONFIG_CRYPTO_AEAD: enabled (as module)
      - CONFIG_CRYPTO_GCM: enabled (as module)
      - CONFIG_CRYPTO_SEQIV: enabled (as module)
      - CONFIG_CRYPTO_GHASH: enabled (as module)
      - CONFIG_XFRM: enabled
      - CONFIG_XFRM_USER: enabled (as module)
      - CONFIG_XFRM_ALGO: enabled (as module)
      - CONFIG_INET_ESP: enabled (as module)
      - CONFIG_INET_XFRM_MODE_TRANSPORT: enabled (as module)
  - "ipvlan":
    - CONFIG_IPVLAN: enabled (as module)
  - "macvlan":
    - CONFIG_MACVLAN: enabled (as module)
    - CONFIG_DUMMY: enabled (as module)
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: missing
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled (as module)
    - CONFIG_BTRFS_FS_POSIX_ACL: enabled
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled (as module)
    - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000

```

**Comment by thaJeztah** [comment]
Thanks @Renay, nothing odd in there

**Comment by thaJeztah** [comment]
Actually, this may already be resolved, see https://github.com/docker/docker/issues/27011 and https://github.com/docker/docker/pull/27467, which will be included in 1.13

/cc @tonistiigi

**Comment by Renay** [comment]
Will check the stackstrace next time to see if this is related

I'm closing for now

Thanks guys ( and sorry for the duplicate issue, didn't see the first one )

**Comment by thaJeztah** [comment]
@Renay no worries! We initially thought it was fixed in 1.12.3, but this issue was already reported in docker 1.11 (so not a regression in 1.12), and the change-set was considered too big for a patch release.
